File: Problem4_Hashset.cpp
// Time Complexity :O(N/k) N=all possible numbers. k=number of predefined buckets
// Space Complexity :O(N+M) M=number of unique values inserted into the bucket
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : Learnt it for the first time. 

//How do I make sure I do not forget them?


class MyHashSet {
    int numBuckets;
    vector<vector<int>> buckets;
    
    //hashing function
    int hashing_function(int key){
        return key%numBuckets;
    }
    
public:
    /** Initialize your data structure here. */
    MyHashSet() {
        numBuckets = 15000;
        buckets = vector<vector<int>>(numBuckets, vector<int>{});
    }
    
    void add(int key) {
        int i = hashing_function(key);
        if(find(buckets[i].begin(), buckets[i].end(),key)==buckets[i].end()){
            buckets[i].push_back(key);
        }
    }
    
    void remove(int key) {
        int i = hashing_function(key);
        auto it = find(buckets[i].begin(), buckets[i].end(),key);
        if(it!=buckets[i].end()){
            buckets[i].erase(it);
        }
    }
    
    /** Returns true if this set contains the specified element */
    bool contains(int key) {
        int i = hashing_function(key);
        auto it = find(buckets[i].begin(), buckets[i].end(),key);
        if(it!=buckets[i].end()){
            return true;
        }
        else{
            return false;
        }
    }
};

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet* obj = new MyHashSet();
 * obj->add(key);
 * obj->remove(key);
 * bool param_3 = obj->contains(key);
 */

File: Problem4_HashSet.java
// Time Complexity : O(n/k) -> n: number of possible keys, k: number of predefined buckets
// Space Complexity : O(k + m) -> m: number of unique keys
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Your code here along with comments explaining your approach
class MyHashSet {
    boolean[][] storage;
    int bucket;
    int bucketItems;

    /** Initialize your data structure here. */
    public MyHashSet() {
        this.bucket = 1000;
        this.bucketItems = 1000;
        this.storage = new boolean[bucket][];
    }
    
    private int bucket(int key) {
        return key % bucket;
    }
    
    private int bucketItems(int key) {
        return key / bucketItems;
    }
    
    public void add(int key) {
        int bucket = bucket(key);
        int bucketItem = bucketItems(key);
        if(storage[bucket] == null) {
            if(bucket == 0) {
                storage[bucket] = new boolean[bucketItems + 1];
            } else {
                storage[bucket] = new boolean[bucketItems];
            }
        }
        storage[bucket][bucketItem] = true;
    }
    
    public void remove(int key) {
        int bucket = bucket(key);
        int bucketItem = bucketItems(key);
        if(storage[bucket] == null)
            return;
        storage[bucket][bucketItem] = false;
    }
    
    /** Returns true if this set contains the specified element */
    public boolean contains(int key) {
        int bucket = bucket(key);
        int bucketItem = bucketItems(key);
        if(storage[bucket] == null)
            return false;
        return storage[bucket][bucketItem];
    }
}